File: gmapsz/src/org/zkoss/gmaps/Gcircle.java
/* Gcircle.java
{{IS_NOTE
	Purpose:
		
	Description:
		
	History:
		Fri Mar 2, 2012 10:38:11 AM, Created by benbai
}}IS_NOTE

Copyright (C) 2012 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under GPL Version 2.0 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
*/

package org.zkoss.gmaps;

import java.util.Map;

import org.zkoss.gmaps.event.CenterChangeEvent;
import org.zkoss.lang.Objects;
import org.zkoss.zk.ui.UiException;
import org.zkoss.zk.ui.event.Event;
import org.zkoss.zk.ui.event.Events;
import org.zkoss.zul.impl.XulElement;
/**
 * Google Maps support Gcircle.
 * 
 * @author benbai
 * @since 3.0.0
 */
public class Gcircle extends XulElement implements Mapitem {
	private static final long serialVersionUID = 7919622962923909687L;
	private LatLng _center = new LatLng(0, 0);
	private boolean _clickable = true; // to match google map default value
	private boolean _editable;
	private String _fillColor;
	private int _fillOpacity = -1; // -1 denotes default, not render it
	private double _radius = 200; // default radius
	private String _strokeColor;
	private int _strokeOpacity = -1; // -1 denotes default, not render it
	private int _strokeWeight = -1; // -1 denotes default, not render it
	private boolean _circleVisible = true; // to match google map default value
	private int _circleZIndex = -1; // -1 denotes default, not render it
	
	/** get the latitude of the center of the circle.
	 * @return double the latitude of the center of the circle.
	 */
	public double getLat () {
		return _center.getLatitude();
	}
	/** set the latitude of the center of the circle.
	 * @param lat the latitude of the center of the circle.
	 */
	public void setLat (double lat) {
		if (!Objects.equals(lat, _center.getLatitude())) {
			setCenter(new LatLng(lat, _center.getLongitude()));
		}
	}

	/** get the longitude of the center of the circle.
	 * @return double the longitude of the center of the circle.
	 */
	public double getLng () {
		return _center.getLongitude();
	}
	/** set the longitude of the center of the circle.
	 * @param lng the longitude of the center of the circle.
	 */
	public void setLng (double lng) {
		if (!Objects.equals(lng, _center.getLongitude())) {
			setCenter(new LatLng(_center.getLatitude(), lng));
		}
	}
	
	/** Sets the Circle's center.
	 * @param center the center of the circle.
	 * @Since 3.0.2
	 */
	public void setCenter(LatLng center) {
		if (!_center.equals(center)) {
			_center = center;
			smartUpdate("center", GmapsUtil.latLngToArray(center));
		}
	}
	
	public LatLng getCenter() {
		return _center;
	}

	/** Returns whether this Gcircle is clickable.
	 * @return whether this Gcircle is clickable.
	 */
	public boolean isClickable () {
		return _clickable;
	}
	/** Sets whether this Gcircle is clickable; default to true.
	 * @param clickable whether enable click the Gcircle.
	 */
	public void setClickable (boolean clickable) {
		if (_clickable != clickable) {
			_clickable = clickable;
			smartUpdate("clickable", _clickable);
		}
	}

	/** Returns whether can edit the Gcircle by mouse; default to false.
	 * @return whether can edit the Gcircle by mouse; default to false.
	 */
	public boolean isEditable () {
		return _editable;
	}
	/** Sets whether can edit the Gcircle by mouse; default to false.
	 * @param editable whether enable edit the Gcircle by mouse.
	 */
	public void setEditable (boolean editable) {
		if (_editable != editable) {
			_editable = editable;
			smartUpdate("editable", _editable);
		}
	}

	/** Returns the fill color of this Gcircle.
	 * @return String the fill color of this Gcircle.
	 */
	public String getFillColor () {
		return _fillColor;
	}
	/** Sets the fill color of this Gcircle.
	 * @param fillColor the fill color of this Gcircle.
	 */
	public void setFillColor (String fillColor) {
		if (!Objects.equals(_fillColor, fillColor)) {
			_fillColor = fillColor;
			smartUpdate("fillColor", _fillColor);
		}
	}

	/** Returns the opacity of the fill color of this Gcircle.
	 * @return int the opacity of the fill color of this Gcircle.
	 */
	public int getFillOpacity() {
		return _fillOpacity;
	}
	/** Sets the opacity of the fill color of this Gcircle.
	 * @param fillOpacity the opacity of the fill color of this Gcircle,
	 * should between 0 and 100.
	 */
	public void setFillOpacity(int fillOpacity) {
		if (fillOpacity < 0 || fillOpacity > 100) {
			throw new UiException("Fill opacity must between 0 to 100 (inclusive): "+fillOpacity);
		}
		if (_fillOpacity != fillOpacity) {
			_fillOpacity = fillOpacity;
			smartUpdate("fillOpacity", _fillOpacity);
		}
	}

	/** Returns the radius of this Gcircle.
	 * @return double the radius of this Gcircle.
	 */
	public double getRadius () {
		return _radius;
	}
	/** Sets the radius of this Gcircle; default to 200.
	 * @param radius the radius of this Gcircle.
	 */
	public void setRadius (double radius) {
		if (radius < 0) {
			throw new UiException("radius must greater then or equal to 0: "+radius);
		}
		if (_radius != radius) {
			_radius = radius;
			smartUpdate("radius", _radius);
		}
	}

	/** Returns the stroke color of this Gcircle.
	 * @return String the stroke color of this Gcircle.
	 */
	public String getStrokeColor () {
		return _strokeColor;
	}
	/** Sets the stroke color of this Gcircle.
	 * @param strokeColor the stroke color of this Gcircle.
	 */
	public void setStrokeColor (String strokeColor) {
		if (!Objects.equals(_strokeColor, strokeColor)) {
			_strokeColor = strokeColor;
			smartUpdate("strokeColor", _strokeColor);
		}
	}

	/** Returns the stroke opacity of this Gcircle.
	 * @return int the stroke opacity of this Gcircle.
	 */
	public int getStrokeOpacity() {
		return _strokeOpacity;
	}
	/** Sets the stroke opacity of this Gcircle.
	 * @param strokeOpacity the stroke opacity of this Gcircle,
	 * should between 0 and 100.
	 */
	public void setStrokeOpacity(int strokeOpacity) {
		if (strokeOpacity < 0 || strokeOpacity > 100) {
			throw new UiException("Stroke opacity must between 0 to 100 (inclusive): "+strokeOpacity);
		}
		if (_strokeOpacity != strokeOpacity) {
			_strokeOpacity = strokeOpacity;
			smartUpdate("strokeOpacity", _strokeOpacity);
		}
	}

	/** Returns the stroke weight of this Gcircle.
	 * @return int the stroke weight of this Gcircle.
	 */
	public int getStrokeWeight() {
		return _strokeWeight;
	}
	/** Sets the stroke weight of this Gcircle.
	 * @param strokeWeight the stroke weight of this Gcircle.
	 */
	public void setStrokeWeight(int strokeWeight) {
		if (strokeWeight < 0) {
			throw new UiException("strokeWeight must greater then or equal to 0: "+strokeWeight);
		}
		if (_strokeWeight != strokeWeight) {
			_strokeWeight = strokeWeight;
			smartUpdate("strokeWeight", _strokeWeight);
		}
	}

	/** Returns whether this Gcircle is visible.
	 * @return boolean whether this Gcircle is visible.
	 */
	public boolean isCircleVisible () {
		return _circleVisible;
	}
	/** Sets whether this Gcircle is visible; default to true.
	 * @param circleVisible whether this Gcircle is visible.
	 */
	public void setCircleVisible (boolean circleVisible) {
		if (_circleVisible != circleVisible) {
			_circleVisible = circleVisible;
			smartUpdate("circleVisible", _circleVisible);
		}
	}

	/** Returns the zIndex of this Gcircle.
	 * @return int the zIndex of this Gcircle.
	 */
	public int getCircleZIndex() {
		return _circleZIndex;
	}
	/** Sets the zIndex of this Gcircle.
	 * @param zIndex the zIndex of this Gcircle.
	 */
	public void setCircleZIndex(int circleZIndex) {
		if (circleZIndex < 0) {
			throw new UiException("circleZIndex must greater then or equal to 0: "+circleZIndex);
		}
		if (_circleZIndex != circleZIndex) {
			_circleZIndex = circleZIndex;
			smartUpdate("circleZIndex", _circleZIndex);
		}
	}

	//super
	public boolean isChildable() {
		return false;
	}
	
	protected void renderProperties(org.zkoss.zk.ui.sys.ContentRenderer renderer)
	throws java.io.IOException {
		super.renderProperties(renderer);

		renderer.render("center",  GmapsUtil.latLngToArray(getCenter()));
		if (!_clickable)
			renderer.render("clickable", _clickable);
		render(renderer, "editable", _editable);
		render(renderer, "fillColor", _fillColor);
		if (_fillOpacity > -1)
			renderer.render("fillOpacity", _fillOpacity);
		if (_radius > -1)
			renderer.render("radius", _radius);
		render(renderer, "strokeColor", _strokeColor);
		if (_strokeOpacity > -1)
			renderer.render("strokeOpacity", _strokeOpacity);
		if (_radius != 200)
			renderer.render("strokeWeight", _strokeWeight);
		if (!_circleVisible)
			renderer.render("circleVisible", _circleVisible);
		if (_circleZIndex > -1)
			renderer.render("circleZIndex", _circleZIndex);
	}
	/** Processes an AU request.
	 *
	 * <p>Default: in addition to what are handled by {@link XulElement#service},
	 * it also handles onSelect.
	 * @since 5.0.0
	 */
	public void service(org.zkoss.zk.au.AuRequest request, boolean everError) {
		final String cmd = request.getCommand();
		if (cmd.equals("onCenterChange")) {
			final CenterChangeEvent evt = CenterChangeEvent.getCenterChangeEvent(request);
			_center = evt.getCenter();
			Events.postEvent(evt);
		} else if (cmd.equals("onRadiusChange")) {
			final Map data = request.getData();
			_radius = ((Number)data.get("radius")).doubleValue();
			Events.postEvent(new Event("onRadiusChange", this, data));
		} else
			super.service(request, everError);
	}
	//register the Gcircle related event
	static {
		addClientEvent(Gcircle.class, "onCenterChange", CE_DUPLICATE_IGNORE);
		addClientEvent(Gcircle.class, "onRadiusChange", CE_DUPLICATE_IGNORE);
	}
}

File: 3.0.0/gmapsz/src/archive/web/js/gmaps/Gcircle.js
/* Gcircle.js

{{IS_NOTE
	Purpose:
		
	Description:
		
	History:
		Fri Mar 02 10:20:11     2012, Created by benbai
		
	Used with ZK 5.0 and later
}}IS_NOTE

Copyright (C) 2012 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under GPL Version 2.0 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
*/
/**
 * Circle that can be overlay on the Google Maps.
 * @since 3.0.0
 */
gmaps.Gcircle = zk.$extends(gmaps.Goverlay, {
	_clickable: true,
	_radius: 200,
	_circleVisible: true,
	$define: {
		/** Returns the Circle's center in double[] array where [0] is lat and [1] is lng;
		 */
		/** Sets the Circle's center in double[] array where [0] is lat and [1] is lng;
		 */
		center: function (c) {
			var circle = this.mapitem_;
			if (circle) {
				// not send event if not changed by drag
				this._centerChangeByServer = true;
				circle.setCenter(new google.maps.LatLng(c[0],c[1]));
				delete this._centerChangeByServer;
			}
		},
		/** Returns whether this Gcircle is clickable.
		 * @return whether this Gcircle is clickable.
		 */
		/** Sets whether this Gcircle is clickable; default to true.
		 * @param b whether enable click the Gcircle.
		 */
		clickable: function (b) {
			this._updateCircle ('clickable', b);
		},
		/** Returns whether can edit the Gcircle by mouse; default to false.
		 * @return whether can edit the Gcircle by mouse; default to false.
		 */
		/** Sets whether can edit the Gcircle by mouse; default to false.
		 * @param b whether enable edit the Gcircle by mouse.
		 */
		editable: function (b) {
			var circle = this.mapitem_;
			if (circle)
				circle.setEditable(b);
		},
		/** Returns the fill color of this Gcircle.
		 * @return String the fill color of this Gcircle.
		 */
		/** Sets the fill color of this Gcircle.
		 * @param v the fill color of this Gcircle.
		 */
		fillColor: function (v) {
			this._updateCircle ('fillColor', v);
		},
		/** Returns the opacity of the fill color of this Gcircle.
		 * @return int the opacity of the fill color of this Gcircle.
		 */
		/** Sets the opacity of the fill color of this Gcircle.
		 * @param v the opacity of the fill color of this Gcircle,
		 * should between 0 and 100.
		 */
		fillOpacity: function (v) {
			this._updateCircle ('fillOpacity', v/100);
		},
		/** Returns the radius of this Gcircle.
		 * @return double the radius of this Gcircle.
		 */
		/** Sets the radius of this Gcircle; default to 200.
		 * @param v the radius of this Gcircle.
		 */
		radius: function (v) {
			var circle = this.mapitem_;
			if (circle) {
				// not send event if not changed by drag
				this._radiusChangeByServer = true;
				circle.setRadius(v);
				delete this._radiusChangeByServer;
			}
		},
		/** Returns the stroke color of this Gcircle.
		 * @return String the stroke color of this Gcircle.
		 */
		/** Sets the stroke color of this Gcircle.
		 * @param v the stroke color of this Gcircle.
		 */
		strokeColor: function (v) {
			this._updateCircle ('strokeColor', v);
		},
		/** Returns the stroke opacity of this Gcircle.
		 * @return int the stroke opacity of this Gcircle.
		 */
		/** Sets the stroke opacity of this Gcircle.
		 * @param strokeOpacity the stroke opacity of this Gcircle,
		 * should between 0 and 100.
		 */
		strokeOpacity: function (v) {
			this._updateCircle ('strokeOpacity', v/100);
		},
		/** Returns the stroke weight of this Gcircle.
		 * @return int the stroke weight of this Gcircle.
		 */
		/** Sets the stroke weight of this Gcircle.
		 * @param v the stroke weight of this Gcircle.
		 */
		strokeWeight: function (v) {
			this._updateCircle ('strokeWeight', v);
		},
		/** Returns whether this Gcircle is visible.
		 * @return boolean whether this Gcircle is visible.
		 */
		/** Sets whether this Gcircle is visible; default to true.
		 * @param b whether this Gcircle is visible.
		 */
		circleVisible: function (b) {
			var circle = this.mapitem_;
			if (circle)
				circle.setVisible(b);
		},
		/** Returns the zIndex of this Gcircle.
		 * @return int the zIndex of this Gcircle.
		 */
		/** Sets the zIndex of this Gcircle.
		 * @param v the zIndex of this Gcircle.
		 */
		circleZIndex: function (v) {
			this._updateCircle ('zIndex', v);
		}
	},
	initMapitem_: function() {
		var	_center = this._center,
			latlng = new google.maps.LatLng(_center[0], _center[1]),
			gcircle = new google.maps.Circle(this.getCircleOptions());

		gcircle.setRadius(this._radius);
		gcircle.setCenter(latlng);
		gcircle.setVisible(this._circleVisible);
		gcircle.setEditable(this._editable);
		gcircle._wgt = this;
		this.mapitem_ = gcircle;
	},
	getCircleOptions: function (prop, val) {
		var circleOption = this._circleOption;
			
		if (!circleOption) {
			var circleZIndex = this._circleZIndex,
				fillColor = this._fillColor,
				fillOpacity = this._fillOpacity,
				strokeColor = this._strokeColor,
				strokeOpacity = this._strokeOpacity,
				strokeWeight = this._strokeWeight;
			this._circleOption =
				{clickable: this._clickable};
			circleOption = this._circleOption;
			if (circleZIndex)
				circleOption['zIndex'] = circleZIndex;
			if (fillColor)
				circleOption['fillColor'] = fillColor;
			if (fillOpacity)
				circleOption['fillOpacity'] = fillOpacity/100;
			if (strokeColor)
				circleOption['strokeColor'] = strokeColor;
			if (strokeOpacity)
				circleOption['strokeOpacity'] = strokeOpacity/100;
			if (strokeWeight)
				circleOption['strokeWeight'] = strokeWeight;
		}

		// override one property, used by setters
		if (prop)
			circleOption[prop] = val;
		return circleOption;
	},
	_updateCircle: function (prop, val) {
		var circle = this.mapitem_;
		if (circle)
			circle.setOptions(this.getCircleOptions(prop, val));
	},
	_initListeners: function() {
		var circle = this.mapitem_,
			wgt = this,
			gevt = google.maps.event;

		this._centerChanged = gevt.addListener(circle, "center_changed",
				function() {if (!wgt._centerChangeByServer) wgt._doCenterChanged(circle.getCenter());});
		this._radiusChanged = gevt.addListener(circle, "radius_changed",
				function() {if (!wgt._radiusChangeByServer) wgt._doRadiusChanged(circle.getRadius());});
		this._lClick = gevt.addListener(circle, "click",
				function(evt) {wgt._doClick(evt);});
		this._dblClick = gevt.addListener(circle, "dblclick",
				function(evt) {wgt._doDoubleClick(evt);});
		this._rClick = gevt.addListener(circle, "rightclick",
				function(evt) {wgt._doRightClick(evt);});
	},
	_doCenterChanged: function (center) {
		var oldCenter = this._center,
			nlat = center.lat(),
			nlng = center.lng(),
			data = {lat:nlat, lng:nlng, oldlat: oldCenter[0], oldlng: oldCenter[1]};
		// update center after data copied
		oldCenter[0] = nlat;
		oldCenter[1] = nlat;
		this.fire('onCenterChange', data, {});
	},
	_doRadiusChanged: function (radius) {
		var nradius = radius,
			data = {radius: nradius, oldRadius: this._radius};

		// update radius after data copied
		this._radius = radius;
		this.fire('onRadiusChange', data, {});
	},
	_doClick: function (evt) {
		this._fireClickEvent(evt.latLng, 'onClick');
		this.doClick_(new zk.Event(this, 'onClick', {latLng: evt.latLng}))
	},
	_doDoubleClick: function (evt) {
		this._fireClickEvent(evt.latLng, 'onDoubleClick');
	},
	_doRightClick: function (evt) {
		this._fireClickEvent(evt.latLng, 'onRightClick');
	},
	_fireClickEvent: function (latlng, evtnm) {
		var parent = this.parent,
			xy = gmaps.Gmaps.latlngToXY(parent, latlng),
			pageXY = gmaps.Gmaps.xyToPageXY(parent, xy.x, xy.y),
			data = {x:xy.x,y:xy.y,pageX:pageXY[0],pageY:pageXY[1]};
	
		this.fire(evtnm, data, {});
	},
	_clearListeners: function () {
		var gevt = google.maps.event;
		if (this._centerChanged) {
			gevt.removeListener(this._centerChanged);
			this._centerChanged = null;
		}
		if (this._radiusChanged) {
			gevt.removeListener(this._radiusChanged);
			this._radiusChanged = null;
		}
		if (this._lClick) {
			gevt.removeListener(this._lClick);
			this._lClick = null;
		}
		if (this._dblClick) {
			gevt.removeListener(this._dblClick);
			this._dblClick = null;
		}
		if (this._rClick) {
			gevt.removeListener(this._rClick);
			this._rClick = null;
		}
	}
});